### PR TITLE
Composite, Proxy and State Pattern

### PR DESCRIPTION
**Pull request by rahulkmr75**
Composite, Proxy and State Pattern

### CODE REVIEW

File: Rahul/ProxyPattern/inc/AbstractOperation.hpp [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+class AbstractOperation{
+    protected:
+    int starsObserved;
+    std::vector<double> distance;
+    public:
+    AbstractOperation();
+    virtual void compute() = 0;
+    virtual int getCount() = 0;
+    virtual std::vector<double> getDistance() = 0;
```
Comment (khatriyash) [inline Rahul/ProxyPattern/inc/AbstractOperation.hpp:16]: Returning a vector isn't a good idea...lot of copying

File: Rahul/ProxyPattern/test.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#include "AbstractOperation.hpp"
+#include "ProxyAstronomicOperation.hpp"
+#include <iostream>
+
+int main(){
+    AbstractOperationPtr proxy = AbstractOperationPtr( (AbstractOperation *)
```
Comment (khatriyash) [inline Rahul/ProxyPattern/test.cpp:6]: minor thing..I think you don't really need to do manual casting...upcast is implicit

File: Rahul/StatePatternCPP/src/GumballMachine.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#include "GumballMachine.hpp"
+#include "NoQuarterState.hpp"
+#include "HasQuarterState.hpp"
+#include "SoldState.hpp"
+#include "SoldOutState.hpp"
+#include "WinnerState.hpp"
+
+GumballMachine::GumballMachine(int balls){
+    this->gumball = balls;
+    this->noQuarterState = AbstractStatePtr((AbstractState *) new NoQuarterState(GumballMachinePtr(this)) );
```
Comment (khatriyash) [inline Rahul/StatePatternCPP/src/GumballMachine.cpp:10]: GumballMachinePtr(this) --- this is dangerous...read about std::enable_shared_from_this

**Review by khatriyash — APPROVED** [review]

**Review by khatriyash — APPROVED** [review]

**Review by khatriyash — APPROVED** [review]

File: Rahul/ProxyPattern/inc/AbstractOperation.hpp (hunk 1 above)
Comment (rahulkmr75) [reply Rahul/ProxyPattern/inc/AbstractOperation.hpp:16]: understood.

**Pull request merged**